smart-queue/enqueue.js

```diff
@@ -0,0 +1,78 @@
+import * as QueueActionTypes from "./action-types";
+
+function validKey(key) {
+  if (key === null) {
+    console.log("Missing key, every queue action should have a key!");
+    return false;
+  }
+  return true;
+}
+
+function indexOfAction(array, key) {
+  return array.findIndex(
+    ({ meta: { offline: { queue = null } } }) => queue && queue.key === key
+  );
+}
+export function enqueue(array, action) {
+  const { meta: { offline: { queue = null } } } = action;
+  let index, existingAction;
+
+  if (!queue) {
+    return [...array, action];
+  }
+
+  const { method = "UNKNOWN", key = null } = queue;
+  if (!validKey(key)) {
+    return array;
+  }
+
+  index = indexOfAction(array, key);
+
+  switch (method) {
+    case QueueActionTypes.CREATE:
+      if (index !== -1) {
+        array[index] = { ...array[index], ...action };
+        return array;
+      }
+      return [...array, action];
+    case QueueActionTypes.DELETE:
+      if (index !== -1) {
+        existingAction = array[index];
+        if (
+          existingAction.meta.offline.queue.method ===
+            QueueActionTypes.UPDATE ||
+          existingAction.meta.offline.queue.method === QueueActionTypes.CREATE
+        ) {
+          array.splice(index, 1);
+        }
+      }
+      return array;
+    case QueueActionTypes.UPDATE:
+      if (index !== -1) {
+        existingAction = array[index];
+        if (
+          existingAction.meta.offline.queue.method === QueueActionTypes.CREATE
+        ) {
+          array[index] = { ...array[index], ...action };
+        }
```

Comment:
Hmm. I don't understand why you can't have multiple updates on the same key, with each new update replacing the one before it. But this is consistent with the specifications.

Reply:
Yes, as per current specifications, we are restricting the user to update a CREATE action only once and replace it with UPDATE action, which leaves no room to update the already updated action. If we remove this check, the user will have the privilege to update any particular action multiple times and this will also allow the user to UPDATE any READ action with the same key as well.

So, either we can restrict UPDATE over existing CREATE or UPDATE action with the same key or we can allow UPDATE over any action with the same key.

Please let me know your thought on this.

Reply:
I think that UPDATE over an UPDATE action should ultimately be supported, it is not unusual to update a resource multiple times, but that should probably be discussed outside of this PR. So this is fine for now.